### PR TITLE
Add a progress bar to report the progress of the extractor

### PR DESCRIPTION
**Pull request by Chik3r**
Adds a progress bar that implemets IProgress and outputs the progress to the console

### CODE REVIEW

File: Consolation/Common/ProgressBar.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Text;
+using System.Threading;
+
+namespace Consolation.Common
+{
+    public class ProgressBar : IDisposable, IProgress<int>
+    {
+        protected virtual byte NumberOfBlocks => 16;
+        protected virtual TimeSpan AnimationInterval => TimeSpan.FromSeconds(1.0 / 8); // Update 8 times a second
```
Comment (Steviegt6) [inline Consolation/Common/ProgressBar.cs:10]: Maybe make this a public property with a protected setter?

File: Consolation/Common/ProgressBar.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+namespace Consolation.Common
+{
+    public class ProgressBar : IDisposable, IProgress<int>
+    {
+        protected virtual byte NumberOfBlocks => 16;
+        protected virtual TimeSpan AnimationInterval => TimeSpan.FromSeconds(1.0 / 8); // Update 8 times a second 
+        
+        private int? _maxElements;
+        private int _currentElements;
+        private readonly Timer _timer;
```
Comment (Steviegt6) [inline Consolation/Common/ProgressBar.cs:14]: Ditto for these as well (and maybe _disposed as well)? I feel like exposing the getters but keeping the setters would be better.

File: TML.Patcher.Backend/Packing/UnpackRequest.cs [hunk 3]
```diff
@@ -20,12 +20,19 @@ public sealed class UnpackRequest
...
+        /// The first report that is sent contains the total amount of files to extract.
+        /// </summary>
+        public IProgress<int> ProgressReporter { get; }
+
+        public UnpackRequest(DirectoryInfo extractDirectory, string filePath, double threads, IProgress<int> progressReporter)
         {
             ExtractDirectory = extractDirectory;
             FilePath = filePath;
             Threads = threads;
+            ProgressReporter = progressReporter;
```
Comment (Steviegt6) [inline TML.Patcher.Backend/Packing/UnpackRequest.cs:35]: Keep Consolation-specific stuff outside of the backend, pl0x. We could implement a hook or something that the progress bar could use, though.

**Comment by Steviegt6** [comment]
The only one that really matters is the last comment. I want the backend to be independent of a console.

File: TML.Patcher.Frontend/Common/Options/UnpackModOption.cs [hunk 4]
```diff
@@ -21,14 +22,20 @@ public override void Execute()
 
             Stopwatch sw = Stopwatch.StartNew();
 
+            ProgressBar bar = new();
+            bar.Start();
```
Comment (Steviegt6) [inline TML.Patcher.Frontend/Common/Options/UnpackModOption.cs:26]: StartNew builder method?

**Comment by Chik3r** [comment]
`_currentElements` still has a private field since you can't do `ref Property`, but can do `ref field` (needed for Interlocked.Add)

**Comment by Steviegt6** [comment]
👍

**Review by Steviegt6 — APPROVED** [review]

**Comment by Steviegt6** [comment]
Merge when you feel like it.

**Pull request merged**